src/TaskExtension.cpp

```diff
@@ -0,0 +1,107 @@
+#include "TaskExtension.h"
+
+#include <QMenu>
+#include <QRegularExpression>
+
+#include "LinkSyntaxHighlighter.h"
+
+static const char INSERT_TASK_REGEX[] = "^(\\s*)([-\\*])?(\\s)?(\\[[x ]\\])?";
+
+static int getTaskCheckmarkPosUnderCursor(const QTextCursor& cursor) {
+    return LinkSyntaxHighlighter::getTaskCheckmarkPosAt(cursor.block().text(),
+                                                        cursor.positionInBlock());
+}
+
+static void toggleTask(QTextCursor* cursor, int pos) {
+    // First insert the new character after the old one, then delete the old character.
+    // This prevents the main cursor from moving as a side effect from the editing.
+    cursor->beginEditBlock();
+    cursor->setPosition(pos + cursor->block().position());
+    cursor->movePosition(QTextCursor::Right, QTextCursor::KeepAnchor);
+    QString c = cursor->selectedText();
+    cursor->movePosition(QTextCursor::Right);
+    if (c == "x") {
+        cursor->insertText(" ");
+    } else {
+        cursor->insertText("x");
+    }
+    cursor->movePosition(QTextCursor::Left);
+    cursor->deletePreviousChar();
+    cursor->endEditBlock();
+}
+
+TaskExtension::TaskExtension(TextEdit* textEdit)
+        : TextEditExtension(textEdit), mInsertToggleTaskAction(std::make_unique<QAction>()) {
+    mInsertToggleTaskAction->setText(tr("Insert/toggle task"));
+    mInsertToggleTaskAction->setShortcut(Qt::ALT | Qt::Key_T);
+    connect(mInsertToggleTaskAction.get(), &QAction::triggered, this, &TaskExtension::insertTask);
+    mTextEdit->addAction(mInsertToggleTaskAction.get());
+}
+
+void TaskExtension::aboutToShowContextMenu(QMenu* menu, const QPoint& pos) {
+    QTextCursor cursor = mTextEdit->cursorForPosition(pos);
+    int checkmarkPos = getTaskCheckmarkPosUnderCursor(cursor);
+    if (checkmarkPos == -1) {
+        return;
+    }
+    menu->addAction(tr("Toggle task completion"), this, [this, pos, checkmarkPos] {
+        // TODO: why can I not pass the cursor here?
```

Comment:
Are you referring to the cursor declared in line 42? What happens if you add it to the captured variables?

Reply:
I get a

```
nanonote/src/TaskExtension.cpp:50:20: error: invalid conversion from ‘const QTextCursor*’ to ‘QTextCursor*’ [-fpermissive]
```

and I am not sure where the `const` comes from.

Reply:
This code is now gone (since it was the equivalent to the "open link under cursor" context menu entry and is now superseded by the "insert/toggle task" shortcut).